### PR TITLE
Fix and extend the Dynamic Primary Pod NIC Name proposal to support rename

### PR DESCRIPTION
**Pull request by orelmisan**
**What this PR does / why we need it**:

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:
Extends the proposal proposed by #324.

**Checklist**

This checklist is not enforcing, but it's a reminder of items that could be relevant to every PR.
Approvers are expected to review this list.

- [ ] Design: A [design document](https://github.com/kubevirt/community/tree/main/design-proposals) was considered and is present (link) or not required
- [ ] PR: The PR description is expressive enough and will help future contributors
- [ ] Code: [Write code that humans can understand](https://en.wikiquote.org/wiki/Martin_Fowler#code-for-humans) and [Keep it simple](https://en.wikipedia.org/wiki/KISS_principle)
- [ ] Refactor: You have [left the code cleaner than you found it (Boy Scout Rule)](https://learning.oreilly.com/library/view/97-things-every/9780596809515/ch08.html)
- [ ] Upgrade: Impact of this change on upgrade flows was considered and addressed if required
- [ ] Testing: New code requires [new unit tests](https://github.com/kubevirt/kubevirt/blob/main/docs/reviewer-guide.md#when-is-a-pr-good-enough). New features and bug fixes require at least on e2e test
- [ ] Documentation: A [user-guide update](https://github.com/kubevirt/user-guide/) was considered and is present (link) or not required. You want a user-guide update if it's a user facing feature / API change.
- [ ] Community: Announcement to [kubevirt-dev](https://groups.google.com/g/kubevirt-dev/) was considered

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: design-proposals/dynamic-primary-pod-nic-name.md [hunk 1]
```diff
@@ -527,3 +531,95 @@ The primary pod interface name is set by the container runtime (for example cri-
...
+and before the VM starts. This hook will trigger logic to adjust the primary interface if necessary.
+
+#### Pros
+- Provides greater flexibility by allowing the domain to be modified in the target, enabling dynamic adjustments during migrations.
+- Allows for future extensibility, making it easier to handle other changes as needed in the migration process.
+- This suggestion can be deferred and implemented when the need arises, as it does not depend on the source virt-launcher.
+- Integrate nicely with network binding plugins' sidecars.
+
+#### Cons
+- Requires additional integration and potentially development with libvirt to provide the hook point.
```
Comment (vladikr) [inline design-proposals/dynamic-primary-pod-nic-name.md:590]: Also, this has never been tested. I'd expect to see a POC before we commit to this process.

File: design-proposals/dynamic-primary-pod-nic-name.md [hunk 2]
```diff
@@ -527,3 +531,95 @@ The primary pod interface name is set by the container runtime (for example cri-
...
+## Alternatives
+### Alternative 1 - Modifying the domain XML at the source virt-launcher
+
+In this alternative, the domain XML would be modified on the source virt-launcher prior to migration. 
+The source virt-launcher would handle adjustments to the primary interface name. 
+
+The target virt-handler will consider the target primary interface name when executing the network setup flow.
+
+#### Pros
+- Could be implemented entirely within KubeVirt, without the need for additional dependencies.
```
Comment (vladikr) [inline design-proposals/dynamic-primary-pod-nic-name.md:601]: This is already an established an tested process that handles CPU affinity and various disk volume readjustments - for me, this is the preferred way to proceed.

File: design-proposals/dynamic-primary-pod-nic-name.md [hunk 3]
```diff
@@ -527,3 +531,95 @@ The primary pod interface name is set by the container runtime (for example cri-
...
+#### Pros
+- Could be implemented entirely within KubeVirt, without the need for additional dependencies. 
+
+#### Cons
+- Limits flexibility in handling dynamic changes that may arise during migration in future versions.
+- Does not integrate smoothly with network binding plugins, as modifying the domain XML solely on the source may not account for network
+configurations handled by plugins on the target.
+
+
+### Alternative 2 - Keeping the original pod interface name
```
Comment (vladikr) [inline design-proposals/dynamic-primary-pod-nic-name.md:609]: That sounds marvelous. Can this actually be achieved?

Comment (orelmisan) [reply design-proposals/dynamic-primary-pod-nic-name.md:609]: This is actually very complicated in practice, and would require us to:
1. Report the currently used pod interface name / tap name from the source virt-handler and keep it as long as the VMI is running.
2. Add a sync step to wait for the source virt-handler to report it before starting the migration.
3. Consider it in the network setup flows of virt-handler and virt-launcher.

**Review by EdDev — COMMENTED** [review]
@orelmisan , getting back to an old idea you had but now from a slightly different perspective (which is to consider it as fixing the current implementation):

- Fix the existing implementation, so the tap device for the primary network is always tap0 like before. No matter the name of the pod interface name.
- The domain xml will never change, remaining with tap0 .
- Once we need to support this name change on migration, we add to the migration status the new field and use it on the virt-handler to setup the network: [pod-iface-name -- bridge name -- tap0].

WDYT?

File: design-proposals/dynamic-primary-pod-nic-name.md [hunk 4]
```diff
@@ -538,3 +542,109 @@ The primary pod interface name is set by the container runtime (for example cri-
...
+
+The target virt-handler will consider the `VMI.Status.MigrationState.TargetPodPrimaryInterfaceName` when executing
+the network setup flow.
+
+#### Pros
+- The addition of the migrationState field could be deferred until the need arises
+- The domain XML will not be modified in either source or target
+
+#### Cons
+- Does not handle network bindings which do not use tap
```
Comment (EdDev) [inline design-proposals/dynamic-primary-pod-nic-name.md:596]: We do not have any such core binding for the primary network.

Network binding plugins that intend to support such scenarios, will need to solve it themselves for now.
In the future, we do plan to support the mutation or observation of the domain xml on the migration target, giving a chance to touch it per the target needs. But this is futuristic.

File: design-proposals/dynamic-primary-pod-nic-name.md [hunk 5]
```diff
@@ -538,3 +542,109 @@ The primary pod interface name is set by the container runtime (for example cri-
...
+
+## Alternatives
+### Alternative 1 - Adding a hook to modify the domain in the target virt-launcher
```
Comment (EdDev) [inline design-proposals/dynamic-primary-pod-nic-name.md:599]: I think you can drop the counting.

Comment (orelmisan) [reply design-proposals/dynamic-primary-pod-nic-name.md]: Done.

File: design-proposals/dynamic-primary-pod-nic-name.md (hunk 4 above)
Comment (orelmisan) [reply design-proposals/dynamic-primary-pod-nic-name.md]: The con was removed.

File: design-proposals/dynamic-primary-pod-nic-name.md [hunk 6]
```diff
@@ -538,3 +542,109 @@ The primary pod interface name is set by the container runtime (for example cri-
...
+to a naming mismatch. Such a failure could, in turn, lead to an upgrade failure. 
+
+## Solution
+
+### Inferring the target pod's primary interface name
+
+All the alternatives listed bellow depend on knowing the target pod's primary interface name.
+However, the current logic in the VMI controller only provides the primary interface name of the source.
+
+To address this, an API change is proposed to enable the reporting if the target pod's primary interface name, allowing
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply design-proposals/dynamic-primary-pod-nic-name.md]: Thanks, done.

File: design-proposals/dynamic-primary-pod-nic-name.md [hunk 7]
```diff
@@ -269,6 +269,17 @@ status:
...
 Alternative solutions could be found in [Appendix A](#appendix-a---alternative-solutions).
 
+### Setting a constant tap name for the VM's primary interface
+
+Currently, for bindings on which KubeVirt is responsible for creating a tap, the tap's name is 
+derived from the pod interface's name. This includes the following bindings:
+- Bridge
+- Masquerade
+- ManagedTap
+
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply design-proposals/dynamic-primary-pod-nic-name.md:276]: An explanation was added.

File: design-proposals/dynamic-primary-pod-nic-name.md [hunk 8]
```diff
@@ -269,6 +269,17 @@ status:
 
 Alternative solutions could be found in [Appendix A](#appendix-a---alternative-solutions).
 
+### Setting a constant tap name for the VM's primary interface
+
+Currently, for bindings on which KubeVirt is responsible for creating a tap, the tap's name is 
+derived from the pod interface's name. This includes the following bindings:
+- Bridge
+- Masquerade
+- ManagedTap
```
[the remark this answers is not in the document]
Comment (orelmisan) [reply design-proposals/dynamic-primary-pod-nic-name.md]: The list was removed.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/community/compare/5e6d35825d35a8f83ecac9f6b118043858f9f345..1ecbe7714e9559a9d53f11b8f63d0804341e99ce): Addressed @EdDev's [comments](https://github.com/kubevirt/community/pull/357#pullrequestreview-2465673639).

**Review by EdDev — APPROVED** [review]
Thanks!

@orelmisan , please note that there are some old comments from @vladikr you may want to address (I think to add points to the pros/cons).

File: design-proposals/dynamic-primary-pod-nic-name.md (hunk 1 above)
Comment (orelmisan) [reply design-proposals/dynamic-primary-pod-nic-name.md:613]: Added this con.

File: design-proposals/dynamic-primary-pod-nic-name.md (hunk 2 above)
Comment (orelmisan) [reply design-proposals/dynamic-primary-pod-nic-name.md]: Added this pro.

**Comment by orelmisan** [comment]
[Change](https://github.com/kubevirt/community/compare/1ecbe7714e9559a9d53f11b8f63d0804341e99ce..97ba1577c1b176801fdac13673407adab4868f24): Addressed @vladikr's comments.

**Comment by vladikr** [comment]
/approve

Thank you @orelmisan and @EdDev ! 
This looks good to me now.

**Review by EdDev — APPROVED** [review]

**Pull request merged**